### ISSUE TITLE
cs: log can freeze Racket

### ISSUE THREAD

**Issue by sorawee** [racket/racket#4167]
**What version of Racket are you using?**
8.4

**What program did you run?**

```
#lang racket

(define v (expt 10 1000000))
(displayln "done")
(time (void (log v 10)))
```

**What should have happened?**

The program should terminate.

**If you got an error message, please include it here.**

N/A

**Please include any other relevant details**

The program doesn't terminate in Racket CS (or at least, doesn't terminate in a reasonable amount of time). 

In Racket BC, the program terminates very quickly:

```
done
cpu time: 162 real time: 161 gc time: 6
```

Moreover, the program can't be ctrl-c and doesn't respond to SIGTERM, etc.

**Comment by samth** [comment]
Also, Gambit uses effectively identical code but completes very quickly. I think the division on this line is the problem: https://github.com/racket/racket/blob/master/racket/src/ChezScheme/s/5_3.ss#L1548

**Comment by lexi-lambda** [comment]
> The same program takes a long time in Chez Scheme, but responds properly to Ctrl-C.

This isn’t right—`log` is not interruptible on Chez Scheme.

Note that sending a single `SIGINT` doesn’t break the program—you have to send it *twice*. Moreover, doing this does something meaningfully different from an ordinary interrupt, as it forcibly resets to the top level. You can see this for yourself by running the following program:

```scheme
(keyboard-interrupt-handler
 (lambda ()
   (display "interrupted\n" (console-output-port))))
(define v (expt 10 100000))
(display "done\n")
(write (log v 10))
```

If you send `SIGINT` just once, then wait for `log` to return, you’ll see the keyboard interrupt handler get executed, then continue with execution (as the keyboard interrupt handler doesn’t escape):

```
$ scheme --script log.scm
done
^Cinterrupted
99999.99999999999
```

But if you send `SIGINT` twice in succession, you’ll still see `interrupted` get printed, but program execution is not allowed to continue:

```
$ scheme --script log.scm
done
^C^Cinterrupted
Noncontinuable interrupt.
```

This behavior is implemented by Chez Scheme’s `SIGINT` handler at the C level here:

https://github.com/racket/racket/blob/fd2fb918c9f80f2f97f141af646c0a10e6b5b066/racket/src/ChezScheme/c/schsig.c#L695-L700

Racket replaces Chez Scheme’s signal handler with its own interrupt handler via rktio, which doesn’t do this. Instead, the keyboard interrupt handler is explicitly invoked by the thread scheduler via the current sandman’s poller procedure.

**Comment by lexi-lambda** [comment]
After a little more investigation, I think Sam is right that the division is where the real denial-of-service issue is coming from. The relevant bit of code is `integer/`, which is what `/` eventually calls:

https://github.com/racket/racket/blob/fd2fb918c9f80f2f97f141af646c0a10e6b5b066/racket/src/ChezScheme/s/5_3.ss#L227-L250

In this case, `big-divide-bignums?` returns `#f`, as both the numerator and denominator have a `$bignum-length` of `103811`, so the final condition fails. This results in a call to `schoolbook-integer/`, which is defined in C and is therefore atomic.

This seems probably undesirable, but I haven’t looked into the linked OpenJDK mailing list discussion to see whether the justification makes sense. If the algorithmic choice is reasonable, it might make sense to reimplement `schoolbook-integer/` in Scheme so that it’s interruptible.

**Comment by mflatt** [comment]
I think the issue may be `gcd`, which is needed by both division algorithms. Chez Scheme has a simple one. BC uses a fancy one from GMP. If that's right, the right change is probably implementing `gcd` in Scheme (as whatever Gambit does?) and using it for bignums just before `schoolbook-integer/`.

**Comment by mflatt** [comment]
Currently, even when deciding to use `quotient&remainder/burnikel-ziegler` for `/`, the implementation starts by using `gcd`... which is Euclidean GCD using schoolbook division. I'm not really sure why this wasn't a bigger problem in the first place.

So, we may not need a fancy `gcd`, and instead just a `gcd` that uses `quotient&remainder/burnikel-ziegler` as appropriate. I have an implementation along those lines, and it makes the example finish quickly, but I'll check more before merging.

**Comment by mflatt** [comment]
Obvious in retrospect: a new `gcd` fixes `(log (expt 10 1000000) 10)` just because it gets powers of two out of the way through shifting, and so `(/ x (ash 1 len)))` ends up looking for the GCD of `x` and `1`.

An example like `(gcd (expt 5 1000000) (expt 3 1000000))` is still relatively slow: 60 seconds on my machine, where BC finishes in 2.3 seconds and Gambit in 1.3 seconds. In that case, enabling `quotient&remainder/burnikel-ziegler` is only a small performance improvement. Still, it's  big improvement overall, because the combination of a Scheme-level `gcd` and using `quotient&remainder/burnikel-ziegler` means that we don't get stuck in the kernel and freeze Racket.

So, the change I'll push fixes the immediate problem, but there's still lots of room for improvement with a better GCD algorithm.
